relay-config/src/redis.rs

```diff
@@ -128,12 +130,98 @@ impl From<RedisConfigFromFile> for RedisConfig {
     }
 }
 
-impl<'de> Deserialize<'de> for RedisConfig {
-    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
-    where
-        D: serde::Deserializer<'de>,
-    {
-        RedisConfigFromFile::deserialize(deserializer).map(Into::into)
+/// Configurations for the various Redis pools used by Relay.
+#[derive(Clone, Debug, Deserialize, Serialize, Eq, PartialEq)]
+#[serde(untagged)]
+pub enum RedisConfigs {
+    /// All pools should be configured the same way.
+    Unified(RedisConfig),
+    /// Individual configurations for each pool.
+    Individual {
+        /// Configuration for the `project_configs` pool.
+        #[serde(skip_serializing_if = "Option::is_none")]
+        project_configs: Option<Box<RedisConfig>>,
+        /// Configuration for the `cardinality` pool.
+        #[serde(skip_serializing_if = "Option::is_none")]
+        cardinality: Option<Box<RedisConfig>>,
+        /// Configuration for the `quotas` pool.
+        #[serde(skip_serializing_if = "Option::is_none")]
+        quotas: Option<Box<RedisConfig>>,
+        /// Configuration for the `misc` pool.
+        #[serde(skip_serializing_if = "Option::is_none")]
+        misc: Option<Box<RedisConfig>>,
+    },
+}
+
+fn create_redis_pool(
+    config: &RedisConfig,
+    cpu_concurrency: usize,
+) -> Result<RedisPool, RedisError> {
+    let options = RedisConfigOptions {
+        max_connections: config
+            .options
+            .max_connections
+            .unwrap_or(cpu_concurrency as u32 * 2)
+            .min(crate::redis::DEFAULT_MIN_MAX_CONNECTIONS),
+        connection_timeout: config.options.connection_timeout,
+        max_lifetime: config.options.max_lifetime,
+        idle_timeout: config.options.idle_timeout,
+        read_timeout: config.options.read_timeout,
+        write_timeout: config.options.write_timeout,
+    };
+
+    match &config.connection {
+        RedisConnection::Single(server) => RedisPool::single(server, options),
+        RedisConnection::Cluster(servers) => {
+            RedisPool::cluster(servers.iter().map(|s| s.as_str()), options)
+        }
+    }
+}
+
+pub(super) fn create_redis_pools(
+    configs: &RedisConfigs,
+    cpu_concurrency: usize,
+) -> Result<relay_redis::RedisPools, RedisError> {
+    match configs {
+        RedisConfigs::Unified(cfg) => {
+            let pool = create_redis_pool(cfg, cpu_concurrency)?;
+            Ok(RedisPools {
+                project_configs: Some(pool.clone()),
+                cardinality: Some(pool.clone()),
+                quotas: Some(pool.clone()),
+                misc: Some(pool),
+            })
```

Comment:
We could also create each pool individually here, and will have to when we want to differentiate between sync and async pools.